File: rspace++/src/rspace/history/root_repository.rs
use super::instances::radix_history::RadixHistory;
use crate::rspace::errors::RootError;
use crate::rspace::hashing::blake2b256_hash::Blake2b256Hash;
use crate::rspace::history::roots_store::RootsStore;

// See rspace/src/main/scala/coop/rchain/rspace/history/RootRepository.scala
pub struct RootRepository {
    pub roots_store: Box<dyn RootsStore>,
}

impl RootRepository {
    pub fn commit(&self, root: &Blake2b256Hash) -> Result<(), RootError> {
        tracing::debug!("[RootRepository] commit {}", root);
        self.roots_store.record_root(root)
    }

    pub fn current_root(&self) -> Result<Blake2b256Hash, RootError> {
        match self.roots_store.current_root()? {
            None => {
                let empty_root_hash = RadixHistory::empty_root_node_hash();
                tracing::debug!(
                    "[RootRepository] currentRoot: empty store, recording {}",
                    empty_root_hash
                );
                self.roots_store.record_root(&empty_root_hash)?;
                Ok(empty_root_hash)
            }
            Some(root) => {
                tracing::debug!("[RootRepository] currentRoot: {}", root);
                Ok(root)
            }
        }
    }

    pub fn validate_and_set_current_root(&self, root: Blake2b256Hash) -> Result<(), RootError> {
        match self
            .roots_store
            .validate_and_set_current_root(root.clone())?
        {
            Some(_) => {
                tracing::debug!("[RootRepository] validateAndSetCurrentRoot OK: {}", root);
                Ok(())
            }
            None => {
                tracing::error!(
                    "[RootRepository] validateAndSetCurrentRoot FAILED: {} not in roots store",
                    root
                );
                Err(RootError::UnknownRootError(format!("unknown root: {}", root)))
            }
        }
    }
}

File: rspace/src/main/scala/coop/rchain/rspace/history/RootRepository.scala
package coop.rchain.rspace.history

import cats.Applicative
import cats.syntax.all._
import cats.effect.Sync
import coop.rchain.rspace.hashing.Blake2b256Hash
import coop.rchain.rspace.history.History.emptyRootHash
import coop.rchain.shared.Log

import scala.Function._

class RootRepository[F[_]: Sync: Log](
    rootsStore: RootsStore[F]
) {
  private def unknownRootError(root: Blake2b256Hash) =
    new RuntimeException(s"unknown root: ${root.bytes.toHex.take(16)}...")

  def commit(root: Blake2b256Hash): F[Unit] =
    Log[F].debug(s"[RootRepository] commit $root") *>
      rootsStore.recordRoot(root)

  def currentRoot(): F[Blake2b256Hash] =
    rootsStore.currentRoot().flatMap {
      case None =>
        Log[F].debug(
          s"[RootRepository] currentRoot: empty store, recording $emptyRootHash"
        ) *> rootsStore.recordRoot(emptyRootHash).as(emptyRootHash)
      case Some(root) =>
        Log[F].debug(
          s"[RootRepository] currentRoot: $root"
        ) *> Applicative[F].pure(root)
    }

  def validateAndSetCurrentRoot(root: Blake2b256Hash): F[Unit] =
    rootsStore.validateAndSetCurrentRoot(root).flatMap {
      case None =>
        Log[F].error(
          s"[RootRepository] validateAndSetCurrentRoot FAILED: $root not in roots store"
        ) *> Sync[F].raiseError[Unit](unknownRootError(root))
      case Some(_) =>
        Log[F].debug(
          s"[RootRepository] validateAndSetCurrentRoot OK: $root"
        )
    }

}
